src/yb/yql/cql/cqlserver/system_query_cache.cc

```diff
@@ -0,0 +1,186 @@
+  scheduler_ = std::make_unique<yb::rpc::Scheduler>(&pool_->io_service());
+
+  LOG(INFO) << "Created system cache updater.";
+
+  if (FLAGS_cql_update_system_cache_msecs > 0) {
+    if (MonoDelta::FromSeconds(FLAGS_cql_system_cache_stale_secs) <
+        MonoDelta::FromMilliseconds(FLAGS_cql_update_system_cache_msecs)) {
+      LOG(WARNING) << "Stale expiration shorter than update rate.";
+    }
+
+    ScheduleRefreshCache(true /* now */);
+  } else {
+    LOG(WARNING) << "System cache created with nonpositive timeout. Disabling scheduling.";
+  }
+}
+
+SystemQueryCache::~SystemQueryCache() {
+  if (pool_) {
+    scheduler_->Shutdown();
+    pool_->Shutdown();
+    pool_->Join();
+  }
+}
+
+boost::optional<RowsResult::SharedPtr> SystemQueryCache::Lookup(const std::string& query) {
+  if (FLAGS_cql_system_cache_stale_secs > 0 &&
+      GetStaleness() > MonoDelta::FromSeconds(FLAGS_cql_system_cache_stale_secs)) {
+    return boost::none;
+  }
+  const std::lock_guard<std::mutex> l(cache_mutex_);
+
+  const auto it = cache_.find(query);
+  if (it == cache_.end()) {
+    return boost::none;
+  } else {
+    return it->second;
+  }
+}
+
+MonoDelta SystemQueryCache::GetStaleness() {
+  const std::lock_guard<std::mutex> l(cache_mutex_);
+  return MonoTime::Now() - last_updated_;
+}
+
+void SystemQueryCache::RefreshCache() {
+  VLOG(1) << "Refreshing system query cache";
+  std::unordered_map<std::string, RowsResult::SharedPtr> new_cache;
+  for (auto query : SYSTEM_QUERIES) {
+    Status status;
+    ExecutedResult::SharedPtr result;
+    ExecuteSync(query, &status, &result);
+
+    if (status.ok()) {
+      new_cache[query] = std::dynamic_pointer_cast<RowsResult>(result);
+    } else {
+      LOG(WARNING) << "Could not execute statement: " << query;
+      // We don't want to update the cache with no data; instead we'll let the
+      // stale cache persist.
+      ScheduleRefreshCache(false /* now */);
+      return;
+    }
+  }
+
+  {
+    const std::lock_guard<std::mutex> l(cache_mutex_);
+    // I'd like to do `cache_ = new_cache`, but for some reason that's not allowed.
+    cache_.clear();
+    for (auto it = new_cache.begin(); it != new_cache.end(); ++it) {
+      cache_[it->first] = it->second;
+    }
+    last_updated_ = MonoTime::Now();
+  }
+
+  ScheduleRefreshCache(false /* now */);
+}
+
+void SystemQueryCache::ScheduleRefreshCache(bool now) {
+  DCHECK(pool_);
+  DCHECK(scheduler_);
+  VLOG(1) << "Scheduling cache refresh";
+
+  scheduler_->Schedule([this](const Status &s) {
+      if (!s.ok()) {
+        LOG(INFO) << "System cache updater scheduler was shutdown: " << s.ToString();
+        return;
+      }
+      this->RefreshCache();
+      }, std::chrono::seconds(now ? 0 : 2));
+}
+
+void SystemQueryCache::ExecuteSync(const std::string& stmt, Status* status,
+    ExecutedResult::SharedPtr* result_ptr) {
+
+  const auto processor = service_impl_->GetProcessor();
+  if (!processor.ok()) {
+    LOG(ERROR) << "Unable to get CQLProcessor";
```

Comment:
"Unable to get CQLProcessor" -> "Unable to get CQLProcessor for system query cache"